### PR TITLE
Dodrio Berry Picking

### PR DESCRIPTION
**Pull request by PikalaxALT**
One nonmatching which I'm too chicken to attempt

### CODE REVIEW

File: src/dodrio_berry_picking.c [hunk 1]
```diff
@@ -1,870 +1,2437 @@
...
+    /*0x000C*/ u8 ALIGNED(4) unk0C;
+    /*0x0010*/ u8 ALIGNED(4) unk10;
+    /*0x0014*/ u8 ALIGNED(4) unk14;
+    /*0x0018*/ u8 ALIGNED(4) unk18;
+    /*0x001C*/ u8 ALIGNED(4) unk1C;
+    /*0x0020*/ u8 ALIGNED(4) unk20;
+    /*0x0024*/ u8 ALIGNED(4) unk24;
+    /*0x0028*/ u8 ALIGNED(4) multiplayerId;
+    /*0x0029*/ u8 filler_0029[7];
+    /*0x0030*/ u8 ALIGNED(4) unk30;
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: maybe `ALIGNED(8)` make more sense

File: src/dodrio_berry_picking.c [hunk 2]
```diff
@@ -1,870 +1,2437 @@
...
-        .baseBlock = 0x67,
+        if (gUnknown_203F3E0->unk112 > gUnknown_203F3E0->unk114)
+            gUnknown_203F3E0->unk114 = gUnknown_203F3E0->unk112;
+        gUnknown_203F3E0->unk112 = 0;
     }
-};
+}
 
-/*static*/ const struct WindowTemplate sUnknown_8475684 =
+static void
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `JOY_NEW`

Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `JOY_NEW`

Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `JOY_NEW`

Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `JOY_NEW`

File: src/dodrio_berry_picking.c [hunk 3]
```diff
@@ -1,870 +1,2437 @@
...
+    case 5:
+        sub_8154128();
+        sub_815417C();
+        sub_8154438();
+        sub_8153E28();
+        gUnknown_203F3E0->unk0C++;
+        break;
+    case 6:
+        BlendPalettes(0xFFFFFFFF, 0x10, 0x00);
+        BeginNormalPaletteFade(0xFFFFFFFF, 0, 16, 0, 0);
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `RGB_BLACK`

File: src/dodrio_berry_picking.c [hunk 4]
```diff
@@ -1,870 +1,2437 @@
...
+        break;
+    }
+}
+
+static void sub_81514F0(void)
+{
+    switch (gUnknown_203F3E0->unk10)
+    {
+    case 0:
+        BeginNormalPaletteFade(0xFFFFFFFF, 0, 0, 16, 0);
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `RGB_BLACK`

File: src/dodrio_berry_picking.c [hunk 5]
```diff
@@ -1,870 +1,2437 @@
...
+    }
+}
+
+static void sub_815159C(void)
+{
+    switch (gUnknown_203F3E0->unk10)
+    {
+    case 0:
+        sub_81549D4(9);
+        BeginNormalPaletteFade(0xFFFFFFFF, 0, 0, 16, 0);
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `RGB_BLACK`

File: src/dodrio_berry_picking.c [hunk 6]
```diff
@@ -1,870 +1,2437 @@
...
+        gUnknown_203F3E0->unk10++;
+        break;
+    case 4:
+        PlayNewMapMusic(MUS_KINOMIKUI);
+        sub_8154540();
+        gUnknown_203F3E0->unk10++;
+        break;
+    case 5:
+        BlendPalettes(0xFFFFFFFF, 16, 0);
+        BeginNormalPaletteFade(0xFFFFFFFF, 0, 16, 0, 0);
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: `RGB_BLACK`

File: src/dodrio_berry_picking.c [hunk 7]
```diff
@@ -1,870 +1,2437 @@
...
-        {3, 0, 1},
-        {0, 1, 2},
-        {1, 2, 3},
-        {2, 3, 0},
-    },
+        if (gUnknown_203F3E0->unk158[i] == 0)
+            gUnknown_203F3E0->unk158[i] = sub_815A5E8(i);
+    }
+
+    // This loop won't ever run, the seemingly poitnless assingment below is to make the compiler
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: typo

File: src/dodrio_berry_picking.c (hunk 2 above)
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking.c]: It's different, because here they're compared as unsigned values.

File: src/dodrio_berry_picking_2.c [hunk 8]
```diff
@@ -1,23 +1,1617 @@
...
+// Unused
+static void sub_81543C4(u8 spriteId)
+{
+    gSprites[spriteId].pos1.x = 20 * spriteId + 50;
+    gSprites[spriteId].pos1.y = 50;
+}
+
+// Gamefreak made a mistake there and goes out of bounds for the data array as it holds 8 elements
+// in turn overwriting sprite's subpriority and subsprites fields.
+#if defined(NONMATCHING)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: probably we can make another macro for known bug fixes, as I feel NONMATCHING is not serving for the purpose

File: src/dodrio_berry_picking_2.c [hunk 9]
```diff
@@ -1,23 +1,1617 @@
...
+static const s16 gUnknown_8478E0E[][2] = {{230, 55}, {30, 74}};
+
+void sub_8154438(void)
+{
+    u8 i;
+    void *ptr = AllocZeroed(0x400);
+    struct SpritePalette sprPal = {sDodrioBerryPlatformPal, 6};
+
+    LZ77UnCompWram(sDodrioBerryPlatformGfx, ptr);
+    if (ptr != NULL)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: Comment `// This check should be one line up.` should be made here

File: src/dodrio_berry_picking_2.c [hunk 10]
```diff
@@ -1,23 +1,1617 @@
...
+    SE_CARD3
+};
+
+void sub_8154128(void)
+{
+    void *ptr = AllocZeroed(0x480);
+    struct SpritePalette sprPal = {sDodrioBerrySpritesPal, 3};
+
+    LZ77UnCompWram(sDodrioBerrySpritesGfx, ptr);
+    if (ptr != NULL)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: Comment `// This check should be one line up.` should be made here

File: src/dodrio_berry_picking_2.c [hunk 11]
```diff
@@ -1,23 +1,1617 @@
...
+        y = gUnknown_203F440->unk3020;
+        if (y == 0)
+            y = 1;
+        FillWindowPixelBuffer(gUnknown_203F440->unk3008[1], PIXEL_FILL(1));
+        AddTextPrinterParameterized(gUnknown_203F440->unk3008[1], 2, gText_Yes, 8, 2, -1, NULL);
+        AddTextPrinterParameterized(gUnknown_203F440->unk3008[1], 2, gText_No, 8, 16, -1, NULL);
+        AddTextPrinterParameterized(gUnknown_203F440->unk3008[1], 2, gText_SelectorArrow2, 0, y == 1 ? 2 : 16, -1, NULL);
+        CopyWindowToVram(gUnknown_203F440->unk3008[1], 3);
+        // Increment state only if A or B button have been pressed.
+        if (gMain.newKeys & A_BUTTON)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking_2.c [hunk 12]
```diff
@@ -1,23 +1,1617 @@
...
+        {
+            PlaySE(SE_SELECT);
+            if (gUnknown_203F440->unk3020 == 0)
+                gUnknown_203F440->unk3020 = 1;
+            gUnknown_203F440->state++;
+        }
+        else if (gMain.newKeys & (DPAD_UP | DPAD_DOWN))
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking_2.c [hunk 13]
```diff
@@ -1,23 +1,1617 @@
...
+                break;
+            case 1:
+                gUnknown_203F440->unk3020 = 2;
+                break;
+            case 2:
+                gUnknown_203F440->unk3020 = 1;
+                break;
+            }
+        }
+        else if (gMain.newKeys & B_BUTTON)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking_2.c [hunk 14]
```diff
@@ -1,23 +1,1617 @@
...
+        {
+            PutWindowTilemap(gUnknown_203F440->unk3008[0]);
+            PutWindowTilemap(gUnknown_203F440->unk3008[1]);
+        }
+        CopyBgTilemapBufferToVram(0);
+        sub_8154324(FALSE);
+        gUnknown_203F440->state++;
+        break;
+    case 4:
+        if (++gUnknown_203F440->unk301C >= 30 && gMain.newKeys & A_BUTTON)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking_2.c [hunk 15]
```diff
@@ -1,23 +1,1617 @@
...
+        if (!IsDma3ManagerBusyWithBgCopy())
+        {
+            PutWindowTilemap(gUnknown_203F440->unk3008[0]);
+            PutWindowTilemap(gUnknown_203F440->unk3008[1]);
+        }
+        CopyBgTilemapBufferToVram(0);
+        gUnknown_203F440->state++;
+        break;
+    case 8:
+        if (++gUnknown_203F440->unk301C >= 30 && gMain.newKeys & A_BUTTON)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking_2.c [hunk 16]
```diff
@@ -1,23 +1,1617 @@
...
+        {
+            PutWindowTilemap(gUnknown_203F440->unk3008[0]);
+            PutWindowTilemap(gUnknown_203F440->unk3008[1]);
+        }
+        CopyBgTilemapBufferToVram(0);
+        FadeOutAndFadeInNewMapMusic(MUS_WIN_YASEI, 20, 10);
+        gUnknown_203F440->state++;
+        break;
+    case 11:
+        if (++gUnknown_203F440->unk301C >= 30 && gMain.newKeys & A_BUTTON)
```
Comment (jiangzhengwenjz) [inline src/dodrio_berry_picking_2.c]: `JOY_NEW`

File: src/dodrio_berry_picking.c (hunk 7 above)
Comment (PikalaxALT) [reply src/dodrio_berry_picking.c]: lol, that typo is in emerald too. blame egg

File: src/dodrio_berry_picking.c (hunk 2 above)
Comment (PikalaxALT) [reply src/dodrio_berry_picking.c]: github mess up line numbers

**Pull request merged**